source/velocity_boundary_conditions/ascii_data.cc

```diff
@@ -0,0 +1,702 @@
+        point_weight[0] = 1 - point_weight[0];
+        surf_vel += add_interpolation_point(spatial_index,
+                                            point_weight,
+                                            time_weight);
+
+        // third and fourth point only in 3D
+        if (dim == 3)
+          {
+            spatial_index[1] = std::min(spatial_index[1]+1,npoints[1]-1);
+            point_weight[1] = 1 - point_weight[1];
+            surf_vel += add_interpolation_point(spatial_index,
+                                                point_weight,
+                                                time_weight);
+
+            spatial_index[0] = spatial_index[0]-1;
+            point_weight[0] = 1 - point_weight[0];
+            surf_vel += add_interpolation_point(spatial_index,
+                                                point_weight,
+                                                time_weight);
+          }
+
+        return surf_vel;
+      }
+
+
+      template <int dim> template <int in, int out>
+      Tensor<1,out>
+      AsciiDataLookup<dim>::convert_tensor (const Tensor<1,in> &old_tensor) const
+      {
+        Tensor<1,out> new_tensor;
+        for (unsigned int i = 0; i < out; i++)
+          if (i < in) new_tensor[i] = old_tensor[i];
+          else new_tensor[i] = 0.0;
+
+        return new_tensor;
+      }
+
+
+      template <int dim>
+      Tensor<1,dim>
+      AsciiDataLookup<dim>::add_interpolation_point(const dealii::TableIndices<dim-1> spatial_index,
+                                                    const double point_weight[dim-1],
+                                                    const double time_weight) const
+      {
+        const Tensor<1,dim> velocity = convert_tensor<3,dim>(velocity_vals(spatial_index));
+        const Tensor<1,dim> old_velocity = convert_tensor<3,dim>(old_velocity_vals(spatial_index));
+
+        Tensor<1,dim> surf_vel = time_weight * velocity + (1-time_weight) * old_velocity;
+
+        surf_vel *= point_weight[0];
+        if (dim == 3)
+          surf_vel *= point_weight[1];
+
+        return surf_vel;
+      }
+
+
+      template <int dim>
+      dealii::TableIndices<dim-1>
+      AsciiDataLookup<dim>::calculate_spatial_index(const Point<dim> &position) const
+      {
+        dealii:TableIndices<dim-1> spatial_index;
+        spatial_index[0] = std::min(static_cast<unsigned int> (position[boundary_dimensions[0]]/spacing[0]),npoints[0]-1);
+        if (dim == 3)
+          spatial_index[1] = std::min(static_cast<unsigned int> (position[boundary_dimensions[1]]/spacing[1]),npoints[1]-1);
+        return spatial_index;
+      }
+
+    template <int dim>
+    void
+    AsciiDataLookup<dim>::calculate_point_weight(double *point_weight,
+                                                 const Point<dim> &position,
+                                                 const dealii::TableIndices<dim-1> spatial_index) const
```

Comment:
Why not have the function return an object of type `std_cxx11::array<dim,double>`? This way you can return an array as, I assume, you really want to :-)